src/snapshots/occlum/unionfs.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+// This unionfs file is used for occlum only
+
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::AtomicUsize;
+
+use anyhow::{anyhow, Result};
+use dircpy::CopyBuilder;
+use fs_extra;
+use fs_extra::dir;
+use nix::mount::MsFlags;
+
+use crate::snapshots::{MountPoint, Snapshotter};
+
+#[derive(Debug)]
+pub struct Unionfs {
+    pub data_dir: PathBuf,
+    pub index: AtomicUsize,
+}
+
+fn clear_path(mount_path: &Path) -> Result<()> {
+    let mut from_paths = Vec::new();
+    let paths = fs::read_dir(
+        mount_path
+            .to_str()
+            .ok_or(anyhow!("mount_path does not exist"))?,
+    )?;
+    for path in paths {
+        from_paths.push(path?.path());
+    }
+    fs_extra::remove_items(&from_paths)?;
+
+    Ok(())
+}
+
+fn create_dir(create_path: &PathBuf) -> Result<()> {
+    if !create_path.exists() {
+        fs::create_dir_all(create_path.as_path())?;
+    }
+
+    Ok(())
+}
+
+fn create_environment(mount_path: &Path) -> Result<()> {
+    let mut from_paths = Vec::new();
+    let mut copy_options = dir::CopyOptions::new();
+    copy_options.overwrite = true;
+
+    #[cfg(target_arch = "x86_64")]
+    {
+        // copy the libs required by occlum to the mount path
+        from_paths.push(PathBuf::from("/lib64/ld-linux-x86-64.so.2"));
+        let mount_path_lib64 = mount_path;
+        let path_lib64 = mount_path_lib64.join("lib64");
+        create_dir(&path_lib64)?;
+
+        fs_extra::copy_items(&from_paths, &path_lib64, &copy_options)?;
+        from_paths.clear();
+    }
+
+    let mount_path_opt = mount_path;
+    let path_opt = mount_path_opt
+        .join("opt")
+        .join("occlum")
+        .join("glibc")
+        .join("lib");
+    fs::create_dir_all(&path_opt)?;
+
+    let occlum_lib = vec![
+        "libc.so.6",
+        "libdl.so.2",
+        "libm.so.6",
+        "libpthread.so.0",
+        "libresolv.so.2",
+        "librt.so.1",
+    ];
+    let ori_occlum_lib_path = Path::new("/opt/occlum/glibc/lib");
+    for lib in occlum_lib.iter() {
+        from_paths.push(ori_occlum_lib_path.join(lib));
+    }
+    fs_extra::copy_items(&from_paths, &path_opt, &copy_options)?;
+    from_paths.clear();
```

Comment:
Currently, `shim-rune` combines Occlum parts with the app using an overlayfs mount. It sounds this copying is redundant.

Reply:
From Occlum pespective, the default glibc libraries in the app container layers couldn't be directly used in Occlum enviroment for some reason. Instead, Occlum provides its own modified glibc libraries. Thus, to generate a runnable RootFS in Occlum, the original glibc libraries need to be replaced with the ones Occlum provided.

Reply:
> Thus, to generate a runnable RootFS in Occlum, the original glibc libraries need to be replaced with the ones Occlum provided.

Do they really need to be replaced if the libOS/container just sets `LD_LIBRARY_PATH` correctly? Right now Occlum libc is both copied (and encrypted) AND overlay mounted from the host into the container rootfs.

Reply:
> Thus, to generate a runnable RootFS in Occlum, the original glibc libraries need to be replaced with the ones Occlum provided.

The "replaced" in above sentence is not that "physically replaced". The original glibc libraries are still in the final rootfs image. Copying the Occlum glibc libraries to the path "/opt/occlum/glibc/lib" makes the LD could find the correct glibc libraries in the first order. The only "physically replaced" file is "/lib64/ld-linux-x86-64.so.2". Occlum uses this LD to find correct libraries in its defined order. (For glibc, firstly it is path "/opt/occlum/glibc/lib").

Reply:
> Copying the Occlum glibc libraries to the path "/opt/occlum/glibc/lib" makes the LD could find the correct glibc libraries in the first order.

my original comment was why copy since `shim-rune` puts an overlayfs mount for them. Is there another reason why these lines needed?

Reply:
> > Copying the Occlum glibc libraries to the path "/opt/occlum/glibc/lib" makes the LD could find the correct glibc libraries in the first order.
> 
> my original comment was why copy since `shim-rune` puts an overlayfs mount for them. Is there another reason why these lines needed?

@zhiwei-intel-h Can you help answer this question? My undertanding is that the `shim-rune` doesn't have these Occlum glibc libraries, thus the the copy is required in agent enclave stage.

Reply:
Now shim-rune do not combines Occlum libs with the app container using overlayfs mount.
From shim-rune design docs https://github.com/confidential-containers/enclave-cc/blob/main/docs/shim-rune.md#occlum,  
shim-rune overlay the boot instance bundle and encrypted unionfs image to generate the final bundle of the app container.
![image](https://user-images.githubusercontent.com/97502165/183594454-cb3a77cb-ecda-447f-b84e-8b75a7f5bce3.png)

The encrypted unionfs image includes the pulled image and some libs required by Occlum.
![image](https://user-images.githubusercontent.com/97502165/183596427-78339e32-0129-4531-a6e3-a7457ed82c95.png)

The boot instance bundle mainly contains Occlum boot template instance.

When launching the app container, the boot instance bundle will find the path of the encrypted unionfs image , 
mount  app image and boot application.

Reply:
> The encrypted unionfs image includes the pulled image and some libs required by Occlum.

@zhiwei-intel-h OK, thanks for the clarification! The way I've used gramine included both "boot instance" and the libs in the same mount so I assumed the same here too. Would this work with Occlum too? As I commented, copying and encrypting these onto the unionfs image sounds inefficient.

Reply:
> > The encrypted unionfs image includes the pulled image and some libs required by Occlum.
> 
> @zhiwei-intel-h OK, thanks for the clarification! The way I've used gramine included both "boot instance" and the libs in the same mount so I assumed the same here too. Would this work with Occlum too? As I commented, copying and encrypting these onto the unionfs image sounds inefficient.

As I know, currently  there is no way for Occlum to simply combine glibc libraries from Occlum through shim-rune in the stage of running app enclave.

Reply:
I would have imagined this is possible using something like Occlum's `hostfs` mounts just like you are doing it with `sefs`.

Reply:
@mythi `hostfs` mount is not a secure way.  To keep whole-flow security for the solution, I don't think `hostfs` should be used to mount "occlum glibc libraries". Moreover, besides "occlum glibc libraries", for some container images, "occlum musl-libc libraries" should be used as well. For shim-rune,  it doesn't know which one should be choosed, the "glibc" or "musl-libc". Thus, let the agent-enclave to do the choice makes sense.

Reply:
@qzheng527 OK, maybe `hostfs` is not the right Occlum term I'm using. I mean that ideally the app enclave with Occlum would be able to just verify the integrity of those Occlum glibc libraries to ensure security and still make the app to use them.